utils/log_parser.py

```diff
@@ -0,0 +1,303 @@
+        return len(self.users_) == 0
+
+    def set_input(self, node, resNo):
+        self.inputs_.append((node, resNo))
+
+    def replace_input(self, oldNode, oldResNo, newNode, newResNo):
+        assert (oldNode, oldResNo) in self.inputs_
+        tmp_inputs = [(newNode, newResNo)]
+        for n, rn in self.inputs_:
+            if n != oldNode or rn != oldResNo:
+                tmp_inputs.append((n, rn))
+        self.inputs_ = tmp_inputs
+
+    def set_scope_of_creation(self, creationScopeName):
+        self.creationScopeName_ = creationScopeName
+
+
+class DottyPrinter:
+    def __init__(self):
+        self.vertices_ = []
+        self.edges_ = []
+        self.uniqueVertexMap = {}
+        self.uniqueVertexNo = 0
+        self.colors = ["AliceBlue", "CadetBlue1", "Coral", "DarkOliveGreen1", "DarkSeaGreen1", "GhostWhite", "Khaki1", "LavenderBlush1", "LemonChiffon1",
+                       "LightSkyBlue", "MistyRose1", "MistyRose2", "PaleTurquoise2", "PeachPuff1", "PowderBlue", "Salmon", "Thistle1", "Thistle3", "Wheat1", "Yellow2"]
+
+    def get_unique_vertex_name(self, node):
+        if node not in self.uniqueVertexMap:
+            self.uniqueVertexMap[node] = self.uniqueVertexNo
+            self.uniqueVertexNo += 1
+        return "v" + str(self.uniqueVertexMap[node])
+
+    def dump_label(self, node):
+        labelStr = "{"
+        labelStr += " {<Inputs>Inputs}|"
+        labelStr += "{" + node.get_kind_name() + "\lname: " + \
+            node.get_name() + "}|"
+        labelStr += " {<Outputs>Outputs}"
+        labelStr += "}"
+        return labelStr
+
+    def get_color(self, node):
+        idx = hash(node.get_kind_name()) % len(self.colors)
+        return self.colors[idx]
+
+    def dump_node(self, node):
+        if not node:
+            return
+        nodeStr = self.get_unique_vertex_name(node) + "[\n"
+        nodeStr += "\tlabel = \""
+        nodeStr += self.dump_label(node)
+        nodeStr += "\"\n"
+        nodeStr += "\tshape = \"record\"\n"
+        nodeStr += "\tstyle=\"filled,rounded\"\n"
+        nodeStr += "\tfillcolor="
+        nodeStr += self.get_color(node)
+        nodeStr += "\npenwidth = 2];\n"
+        self.vertices_.append(nodeStr)
+
+    def visitNodes(self):
+        for k, n in NODES_MAP:
+            self.dump_node(NODES_MAP[(k, n)])
+
+    def visitEdges(self):
+        for k, n in NODES_MAP:
+            node = NODES_MAP[(k, n)]
+            for nodeInput in node.get_inputs():
+                i = nodeInput[0]
+                if (i.get_kind_name(), i.get_name()) not in NODES_MAP:
+                    print(i.get_kind_name(), i.get_name())
+                edgeStr = self.get_unique_vertex_name(i)+":Outputs -> "
+                edgeStr += self.get_unique_vertex_name(node)+":Inputs"
+                self.edges_.append(edgeStr)
+
+    def dump_graph(self, dagName):
+        self.visitNodes()
+        self.visitEdges()
+        f = open(dagName + "_dotty.dot", "w")
```

Comment:
Consider using `with` instead of explicitly calling `open` and `close`:
```
with open(f"{dagName}_dotty.dot", "w") as f:
  f.write(...)
  ...
```